Feat(user): add POST /users/create endpoint without validation (#618)

Heyvkram opened the pull request:
This pull request modifies the previous one (https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/946), by removing the ADMIN role requirement. Previously, the POST /itachallenge/api/v1/admin/users/create route required an authenticated token with admin privileges. This validation has been removed to allow any client to register users.

Changes Made
AdminCreateUserController.java:

- Removed the dependency on IJwtService.
 
- Removed all logic related to reading the Authorization header and validating the user's role.
 
- The controller method now directly invokes AdminCreateUserService without any permission checks.

- The class names (Admin...) and endpoint path (.../admin/...) are deliberately kept to minimize the impact of this temporary patch and to facilitate future reversion or refactoring.

✅ Manual Testing (Postman)
The endpoint has been tested manually and is working correctly.

Method: POST

URL: http://localhost:8764/itachallenge/api/v1/admin/users/create

Body:
Select raw and JSON.

JSON
{
    "username": "testOne"
}

Expected Result:

Status: 201 Created

Body: A JSON object containing the new user's uuid, username, and role.

JSON

{
    "uuid": "some-unique-uuid-string",
    "username": "testuser_en",
}

acocinas committed:
refactor: Rename user request DTOs for admin context

acocinas committed:
feat: Add endpoint for admin to create users

acocinas committed:
test: Add unit tests for admin user creation feature

acocinas committed:
test: Add unit tests for JwtServiceImpl to improve coverage

acocinas committed:
chore(test): Remove comments to comply with SonarQube rules

Heyvkram committed:
controller changes to make the endpoint free to role validations

acocinas reviewed (changes requested):
Hi! Thanks for taking on this task.

After reviewing the PR, I think there might be some missing context from the original user story and why the new jwt-core module task was created.

The code you've modified, especially JwtServiceImpl, is based on my previous implementation. That implementation is precisely what we are about to remove and replace entirely with the new jwt-core module, which will centralize all authentication, so I don't think it's necessary to use it in this PR.

The correct strategy should be: to push the endpoint without role validation and wait for the work from the main jwt-core story to be merged into develop.

Heyvkram committed:
delete the JWT files due a future implementation

Heyvkram committed:
Implement AdminCreateUserControllerTest

Heyvkram commented:
Thank you very much for the review!

I have deleted as you say all the JWT files that you implemented in the previous PR.

I have implemented also a test due to Sonar specifications.

acocinas reviewed (changes requested):
Just a suggestion to make this PR even cleaner, now that JwtServiceImpl has been removed:

The UnauthorizedException.java file appears to be unused now. To keep the PR focused on its main goal, maybe we could remove it?

Along the same lines, the handleUnauthorizedException method in UserGlobalähän now seems to be unreachable. It might be a good candidate for removal as well.

acocinas commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/AdminCreateUserControllerTest.java:
"Please remove the comments in the code."

Heyvkram replied:
Done!

Heyvkram committed:
Delete unused exceptions and delete comment code

Heyvkram commented:
Good point, thanks! I have already corrected everything suggested and required!

acocinas commented:
Great work on the changes! From my side, the PR is approved.

The final solution is clean, clear, and perfectly meets the objective of the task.

acocinas reviewed (approved):
Great work on the changes! From my side, the PR is approved.

The final solution is clean, clear, and perfectly meets the objective of the task.

Heyvkram committed:
modify CHANGELOG and envCI.dev

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/controller/AdminCreateUserController.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @PostMapping("/users/create")
+    public Mono<ResponseEntity<AdminCreateUserResponseDto>> createUser(
```
The current naming with the AdminCreateUser* prefix is ​​well-suited if the goal is to restrict this endpoint to ADMIN again soon.
To avoid confusion in the meantime, it would be helpful to leave a comment in the code indicating that the restriction is temporary and is expected to be reverted soon.

Heyvkram replied:
Yes, good point! Comment added!

ivilarop replied:
Great — just one small suggestion: it’s common to prefix temporary comments like this with // TODO: so that they stand out more clearly in code reviews and IDEs.

That way, future contributors can spot that this is something planned to be revisited.

Heyvkram replied:
OK! that's true! done!

ivilarop replied:
Ok thks!

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/dto/AdminCreateUserResponseDto.java:
```diff
@@ -0,0 +1,21 @@
+public class AdminCreateUserResponseDto {
+
+    @JsonProperty("uuid")
+    private String uuid;
```
There is no defined nomenclature but we can adapt it to the naming of the uuids that already exist to be aligned.

Heyvkram replied:
Yes, good point! I change it to:

@JsonProperty(value ="uuid_user")
    private String userId;
   
as the other DTOs

ivilarop replied:
ok thks!

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/dto/AdminCreateUserRequestDto.java:
```diff
@@ -5,7 +5,7 @@
 
 @Getter
 @Setter
-public class CreateUserRequest {
```
I don't see this naming refactoring in the PR description. Why did you decide to use the same DTO for user actions and admin actions?

Heyvkram replied:
Sorry for not including it. The name refactoring was requested in the previous PR, so my colleague made the change in the previous PR.
As for the DTO, we've been looking, and it's only being used in the create user. The login is using entities, so it doesn't affect the login!

ivilarop replied:
Okay, we'll leave it like this for now. A technical issue has arisen due to the refactoring of the login endpoint, and this will require a user ID, hence the question: https://tree.taiga.io/project/luisvi-ita-challenges/issue/631 . We can resolve the conversation.

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/controller/AdminCreateUserController.java:
```diff
@@ -0,0 +1,27 @@
+@RequestMapping("/itachallenge/api/v1/admin")
+public class AdminCreateUserController {
+
+    private final AdminCreateUserService adminCreateUserService;
```
A line break is missing to improve readability, right?

Heyvkram replied:
Totally right! Solved!

ivilarop replied:
ook thks!

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.UUID;
+
+@Service
+public class AdminCreateUserService {
```
is there no missing interface?

Heyvkram replied:
My bad!
I created the interface and implemented all the corrections needed with the service and the controller!

ivilarop replied:
ok thks!Nice move extracting IAdminCreateUserService. This keeps the controller loosely coupled and makes mocking or swapping the service much easier in the future. Solid design choice

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/dto/AdminCreateUserResponseDto.java:
```diff
@@ -0,0 +1,21 @@
+    private String username;
+
+    @JsonProperty("role")
+    private String role;
```
Is this role field worth it if it is for student (user) registration only?If needed, shouldn't it be of the enum type that we have already created?

Heyvkram replied:
It has no use at the momment, deleted!

ivilarop replied:
ok thks!

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java:
```diff
@@ -0,0 +1,44 @@
+                            .map(savedUser -> AdminCreateUserResponseDto.builder()
+                                    .uuid(savedUser.getUuid().toString())
+                                    .username(savedUser.getUsername())
+                                    .role(savedUser.getRole().toString())
```
is toString necessary?

Heyvkram replied:
I think it is necessary in this case, but I may be wrong!

The .userId() expects a String because the DTO format and we are passing it a UUID object via savedUser.getUuid(), so we need the toSting() ""bridge"" to transform that type of object into a Sting.

If I'm wrong and it's another method or for code it has to be done diferently please let me know!

ivilarop replied:
Okay, it wasn't necessary for the role, but since we've removed it, it no longer applies. It's fine.

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java:
```diff
@@ -0,0 +1,44 @@
+
+        return userRepository.findByUsername(username)
+                .flatMap(existingUser ->
+                        Mono.<AdminCreateUserResponseDto>error(new UsernameAlreadyExistsException("Username " + username + " already exists."))
```
Would it be valid to throw new UsernameAlreadyExistsException(username);?

Heyvkram replied:
Yes! my bad! the exception class retuns the message already! the way you suggest is the correct and the cleanest way! Correction done! It was a lapsus!

ivilarop replied:
ok thks!

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java:
```diff
@@ -0,0 +1,44 @@
+        this.userRepository = userRepository;
+    }
+
+    public Mono<AdminCreateUserResponseDto> createUser(AdminCreateUserRequestDto request) {
```
Could we add a logger to have traceability?

Heyvkram replied:
Logger added! Now we can trace the atempt to create the user with name X, get a message of success or a message with the exception in case of it already exists!

Heyvkram replied:
Tested via postman!

ivilarop replied:
Ok perfect! thks

Heyvkram committed:
Code corrections, interface of AdminCreateUserService creation, logger implementation in createUser

Heyvkram committed:
Test code correction

acocinas commented on itachallenge-user/src/test/java/com/itachallenge/user/service/AdminCreateUserServiceTest.java:
Hey, this piece of code is so clear and well-written that I don't think this comment is needed anymore. The code speaks for itself!

Heyvkram replied:
Thanks for the comment!

Heyvkram committed:
comment line deleted

acocinas reviewed (approved):
✅ Overall Review
This PR successfully delivers the core requirement of the task: to provide an endpoint for creating users without role validation, effectively unblocking the frontend team.

Given the context that this is a temporary measure, the approach is understandable. The feedback below focuses on acknowledging the accepted risks and ensuring they are properly tracked.

👍 Highlights (Positive Points)

Solid Service Logic: The AdminCreateUserService is well-implemented. It correctly uses reactive principles (flatMap, switchIfEmpty) to check for existing users before saving a new one.

Good Communication: The code includes comments explaining the temporary nature of the endpoint, and the developer has been responsive to feedback regarding logging.

🏁 Conclusion
Approve. 👍

Heyvkram commented:
Thanks for the review @acocinas  !

isaac-diez reviewed (approved):
✅ Review completed
Thanks for the PR. I've reviewed the code, tests, and previous comments, and here are my comments:

✅ What’s working well
- The /admin/users/create endpoint correctly addresses the functional need described in the subtask, allowing user creation from the frontend without validation.
- The existing username-based validation logic is properly encapsulated and tested.
- Keeping the original /admin/... path makes it easier to restrict access or roll back in the future (#589).
- The design follows good practices: separate DTOs, well-scoped services, informative logs, and complete test coverage.
- Reviewer suggestions have been properly addressed.

🧠 Minor notes
Regarding the DTO name (AdminCreateUserRequestDto), even though it may be a bit misleading right now due to the lack of role validation, keeping it as is makes sense until #589 is implemented.

📌 Conclusion
Great job overall. Approved ✅
Just make sure to resolve any "Pending" comments if applicable.

Heyvkram committed:
correction to the coment in file AdminCreateUserController due to clarify next steps
